utils/zerovec/src/ule/vartuple.rs

```diff
@@ -0,0 +1,201 @@
+// this same safety constraint. Further, there is no padding, because repr(C)
+// does not add padding when all fields are align(1).
+//
+// <https://doc.rust-lang.org/reference/type-layout.html#the-c-representation>
+//
+// Pointers to `VarTupleULE` are fat pointers with metadata equal to the
+// metadata of the inner DST field V.
+//
+// <https://doc.rust-lang.org/stable/std/ptr/trait.Pointee.html>
+//
+// ## Checklist
+//
+// Safety checklist for `VarULE`:
+//
+// 1. align(1): see "Representation" above.
+// 2. No padding: see "Representation" above.
+// 3. `validate_byte_slice` checks length and defers to the inner ULEs.
+// 4. `validate_byte_slice` checks length and defers to the inner ULEs.
+// 5. `from_byte_slice_unchecked` returns a fat pointer to the bytes.
+// 6. All other methods are left at their default impl.
+// 7. The two ULEs have byte equality, so this composition has byte equality.
+unsafe impl<A, V> VarULE for VarTupleULE<A, V>
+where
+    A: AsULE + 'static,
+    V: VarULE + ?Sized,
+{
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), UleError> {
+        // TODO: use split_first_chunk_mut in 1.77
+        if bytes.len() < size_of::<A::ULE>() {
+            return Err(UleError::length::<Self>(bytes.len()));
+        }
+        let (sized_chunk, variable_chunk) = bytes.split_at(size_of::<A::ULE>());
+        A::ULE::validate_byte_slice(sized_chunk)?;
+        V::validate_byte_slice(variable_chunk)?;
+        Ok(())
+    }
+
+    unsafe fn from_byte_slice_unchecked(bytes: &[u8]) -> &Self {
+        // Safety: The DST of VarTupleULE is a pointer to the `sized` element and has a metadata
+        // equal to the metadata of the `variable` field.
+        #[allow(clippy::panic)] // panic is documented in function contract
+        if bytes.len() < size_of::<A::ULE>() {
+            panic!("from_byte_slice_unchecked called with short slice")
+        }
+        let (_sized_chunk, variable_chunk) = bytes.split_at(size_of::<A::ULE>());
+        let variable_ref = V::from_byte_slice_unchecked(variable_chunk);
+        let variable_ptr: *const V = variable_ref;
+        // We should use the pointer metadata APIs here when they are stable: https://github.com/rust-lang/rust/issues/81513
+        // For now we rely on all DST metadata being a usize to extract it via a fake slice pointer
+        // Rust doesn't know that `&V` is a fat pointer so we have to use transmute_copy
+        debug_assert_eq!(size_of::<*const V>(), size_of::<*const [u8]>());
+        let variable_fat_ptr: *const [u8] = core::mem::transmute_copy(&variable_ptr);
+        let (_ptr, metadata): (usize, usize) = core::mem::transmute(variable_fat_ptr);
+        let composed_fat_ptr: *const [u8] = core::slice::from_raw_parts(bytes.as_ptr(), metadata);
+        debug_assert_eq!(size_of::<*const Self>(), size_of::<*const [u8]>());
+        let composed_ref: *const Self = core::mem::transmute_copy(&composed_fat_ptr);
+        &*(composed_ref)
```

Comment:
nit: we should also have comments explaining what's going on with all the transmutes

Reply:
Done